### PR TITLE
Sponsored tx part1

### PR DESCRIPTION
**Pull request by DOBEN**
## Purpose

This is the first part of https://github.com/Concordium/concordium-rust-smart-contracts/pull/318 without the integration tests. 

We split up the linked PR because we have moved the signature/public_key types to `concordium-contracts-common` package already and merged it. We want to get the associated types removed from this package soonish to be in sync with the other packages. This will allow us to continue working without disruption on these packages. The integration tests will be added in a second PR once it is ready and the associated testing libraries are released.

## Changes

This PR https://github.com/Concordium/concordium-rust-smart-contracts/pull/318 without the integration tests.

### CODE REVIEW

**Review by abizjak — APPROVED** [review]

File: examples/cis3-nft-sponsored-txs/src/lib.rs [hunk 1]
```diff
@@ -295,45 +256,45 @@ struct SetImplementorsParams {
...
 #[derive(Serialize, SchemaType)]
 pub struct PermitParam {
-    /// Signature/s. The CIS3 standard supports multi-sig accounts but because
-    /// the `public_key_registry` in this contract maps one public key to one
-    /// account, only one signature has to be provided for this contract. This
-    /// signature has to be at the key 0 in both maps below.
-    #[concordium(size_length = 1)]
-    signature: BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>,
+    /// Signature/s. The CIS3 standard supports multi-sig accounts. But for
+    /// simplicity, this contract only supports signatures that were generated
```
Comment (limemloh) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:279]: Is it still true that we only support accounts with one public key?

**Review by limemloh — APPROVED** [review]

File: examples/cis3-nft-sponsored-txs/src/lib.rs (hunk 1 above)
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:279]: Not true anymore. I updated the comment.

File: examples/cis3-nft-sponsored-txs/src/lib.rs [hunk 2]
```diff
@@ -295,45 +256,45 @@ struct SetImplementorsParams {
...
 pub struct PermitParamPartial {
-    /// Signature/s. The CIS3 standard supports multi-sig accounts but because
-    /// the `public_key_registry` in this contract maps one public key to one
-    /// account, only one signature has to be provided for this contract. This
-    /// signature has to be at the key 0 in both maps below.
-    #[concordium(size_length = 1)]
-    signature: BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>,
+    /// Signature/s. The CIS3 standard supports multi-sig accounts. But for
+    /// simplicity, this contract only supports signatures that were generated
+    /// by accounts with exactly one public key (one credential with one public
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:294]: Not true anymore. I updated the comment.

File: examples/cis3-nft-sponsored-txs/src/lib.rs [hunk 3]
```diff
@@ -1120,76 +1009,104 @@ fn contract_balance_of<S: HasStateApi>(
...
-    pub queries: Vec<PublicKeyQuery>,
+    pub queries: Vec<AccountAddressStruct>,
 }
 
-/// Part of the parameter type for the contract function `publicKeyOf`.
+/// Part of the parameter type for the contract function `publicKeyOf/nonceOf`.
 #[derive(Debug, Serialize, SchemaType)]
-pub struct PublicKeyQuery {
-    /// The account for which the public key and nonce should be queried.
+pub struct AccountAddressStruct {
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:1038]: Simplified it to remove this extra type `AccountAddressStruct` by using `AccountAddress` instead:

```
pub struct VecOfAccountAddresses {
    /// List of queries.
    #[concordium(size_length = 2)]
    pub queries: Vec<AccountAddress>,
}
```

**Pull request merged**